src/main/java/io/netty/incubator/codec/http3/Http3FrameCodec.java

```diff
@@ -721,6 +723,69 @@ private void unsetState(int toUnset) {
         }
     }
 
+    private static final class WriteResumptionListener
+            implements GenericFutureListener<Future<? super QuicStreamChannel>> {
+        private static final Object FLUSH = new Object();
+        private final Deque<Object> buffer;
+        private final ChannelHandlerContext ctx;
+        private final Http3FrameCodec codec;
+
+        WriteResumptionListener(ChannelHandlerContext ctx, Http3FrameCodec codec) {
+            this.ctx = ctx;
+            this.codec = codec;
+            buffer = new ArrayDeque<>(4); // assuming we will buffer header, data, trailer and a flush
+            assert codec.qpackAttributes != null;
+            codec.qpackAttributes.whenEncoderStreamAvailable(this);
+        }
+
+        @Override
+        public void operationComplete(Future<? super QuicStreamChannel> future) {
+            drain();
+        }
+
+        void enqueue(Object msg, ChannelPromise promise) {
+            assert ctx.channel().eventLoop().inEventLoop();
+
+            buffer.addLast(new BufferedEntry(msg, promise));
+        }
+
+        void enqueueFlush() {
+            assert ctx.channel().eventLoop().inEventLoop();
+
+            buffer.addLast(FLUSH);
+        }
+
+        void drain() {
+            assert ctx.channel().eventLoop().inEventLoop();
+            boolean flushSeen = false;
+            for (Object entry = buffer.pollFirst(); entry != null; entry = buffer.pollFirst()) {
+                if (entry == FLUSH) {
+                    flushSeen = true;
+                } else {
+                    assert entry instanceof BufferedEntry;
+                    BufferedEntry bufferedEntry = (BufferedEntry) entry;
+                    codec.write0(ctx, bufferedEntry.msg, bufferedEntry.promise);
+                }
+            }
+            // indicate that writes do not need to be enqueued. As we are on the eventloop, no other writes can happen
+            // while we are draining, hence we would not write out of order.
+            codec.writeResumptionListener = null;
+            if (flushSeen) {
```

Comment:
should this part be added to a finally block ?

Reply:
Good point, yes will do